File: frontend/src/utils/timezoneUtils.js
/**
 * Frontend Timezone Utilities for Hong Kong Time
 * Ensures consistent timezone handling across the entire frontend application
 */

// Hong Kong timezone configuration
const HK_TIMEZONE = 'Asia/Hong_Kong';
const HK_LOCALE = 'en-HK';

/**
 * Get current time in Hong Kong timezone
 * @returns {Date} Current date/time in Hong Kong timezone
 */
export const getHKNow = () => {
  return new Date();
};

/**
 * Get current time in Hong Kong timezone as ISO string
 * @returns {string} Current time in Hong Kong timezone as ISO format
 */
export const getHKNowISO = () => {
  return new Date().toISOString();
};

/**
 * Format a date string to Hong Kong timezone
 * @param {string|Date} dateString - Date string or Date object
 * @param {Object} options - Formatting options
 * @returns {string} Formatted date string in Hong Kong timezone
 */
export const formatHKDate = (dateString, options = {}) => {
  if (!dateString) return '-';
  
  try {
    const date = new Date(dateString);
    
    // Default options for Hong Kong formatting
    const defaultOptions = {
      timeZone: HK_TIMEZONE,
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
      ...options
    };
    
    return date.toLocaleDateString(HK_LOCALE, defaultOptions);
  } catch (error) {
    console.warn('Error formatting date:', error);
    return dateString;
  }
};

/**
 * Format a date string to Hong Kong timezone with time
 * @param {string|Date} dateString - Date string or Date object
 * @param {Object} options - Formatting options
 * @returns {string} Formatted date and time string in Hong Kong timezone
 */
export const formatHKDateTime = (dateString, options = {}) => {
  if (!dateString) return '-';
  
  try {
    const date = new Date(dateString);
    
    // Default options for Hong Kong date/time formatting
    const defaultOptions = {
      timeZone: HK_TIMEZONE,
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit',
      ...options
    };
    
    return date.toLocaleString(HK_LOCALE, defaultOptions);
  } catch (error) {
    console.warn('Error formatting date/time:', error);
    return dateString;
  }
};

/**
 * Format a date string to Hong Kong timezone with time (short format)
 * @param {string|Date} dateString - Date string or Date object
 * @returns {string} Formatted date and time string in Hong Kong timezone (short)
 */
export const formatHKDateTimeShort = (dateString) => {
  if (!dateString) return '-';
  
  try {
    const date = new Date(dateString);
    
    return date.toLocaleString(HK_LOCALE, {
      timeZone: HK_TIMEZONE,
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
      hour: '2-digit',
      minute: '2-digit'
    });
  } catch (error) {
    console.warn('Error formatting date/time (short):', error);
    return dateString;
  }
};

/**
 * Format a date string to Hong Kong timezone (date only)
 * @param {string|Date} dateString - Date string or Date object
 * @returns {string} Formatted date string in Hong Kong timezone (date only)
 */
export const formatHKDateOnly = (dateString) => {
  if (!dateString) return '-';
  
  try {
    const date = new Date(dateString);
    
    return date.toLocaleDateString(HK_LOCALE, {
      timeZone: HK_TIMEZONE,
      year: 'numeric',
      month: '2-digit',
      day: '2-digit'
    });
  } catch (error) {
    console.warn('Error formatting date only:', error);
    return dateString;
  }
};

/**
 * Format a date string to Hong Kong timezone (time only)
 * @param {string|Date} dateString - Date string or Date object
 * @returns {string} Formatted time string in Hong Kong timezone (time only)
 */
export const formatHKTimeOnly = (dateString) => {
  if (!dateString) return '-';
  
  try {
    const date = new Date(dateString);
    
    return date.toLocaleTimeString(HK_LOCALE, {
      timeZone: HK_TIMEZONE,
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit'
    });
  } catch (error) {
    console.warn('Error formatting time only:', error);
    return dateString;
  }
};

/**
 * Get relative time in Hong Kong timezone (e.g., "2 hours ago")
 * @param {string|Date} dateString - Date string or Date object
 * @returns {string} Relative time string
 */
export const getHKRelativeTime = (dateString) => {
  if (!dateString) return '-';
  
  try {
    const date = new Date(dateString);
    const now = new Date();
    const diffMs = now - date;
    const diffMins = Math.floor(diffMs / (1000 * 60));
    const diffHours = Math.floor(diffMs / (1000 * 60 * 60));
    const diffDays = Math.floor(diffMs / (1000 * 60 * 60 * 24));
    
    if (diffMins < 1) return 'Just now';
    if (diffMins < 60) return `${diffMins} minute${diffMins > 1 ? 's' : ''} ago`;
    if (diffHours < 24) return `${diffHours} hour${diffHours > 1 ? 's' : ''} ago`;
    if (diffDays < 7) return `${diffDays} day${diffDays > 1 ? 's' : ''} ago`;
    
    return formatHKDateOnly(dateString);
  } catch (error) {
    console.warn('Error getting relative time:', error);
    return dateString;
  }
};

/**
 * Check if a date is today in Hong Kong timezone
 * @param {string|Date} dateString - Date string or Date object
 * @returns {boolean} True if date is today
 */
export const isHKToday = (dateString) => {
  if (!dateString) return false;
  
  try {
    const date = new Date(dateString);
    const today = new Date();
    
    const dateHK = date.toLocaleDateString(HK_LOCALE, { timeZone: HK_TIMEZONE });
    const todayHK = today.toLocaleDateString(HK_LOCALE, { timeZone: HK_TIMEZONE });
    
    return dateHK === todayHK;
  } catch (error) {
    console.warn('Error checking if date is today:', error);
    return false;
  }
};

/**
 * Get Hong Kong timezone information
 * @returns {Object} Timezone information
 */
export const getHKTimezoneInfo = () => {
  const now = new Date();
  
  return {
    timezone: HK_TIMEZONE,
    locale: HK_LOCALE,
    currentTime: now.toLocaleString(HK_LOCALE, { timeZone: HK_TIMEZONE }),
    currentDate: now.toLocaleDateString(HK_LOCALE, { timeZone: HK_TIMEZONE }),
    utcOffset: now.toLocaleString('en-US', { 
      timeZone: HK_TIMEZONE, 
      timeZoneName: 'shortOffset' 
    }).split(' ').pop()
  };
};

/**
 * Convert a date to Hong Kong timezone for comparison
 * @param {string|Date} dateString - Date string or Date object
 * @returns {Date} Date object in Hong Kong timezone
 */
export const toHKDate = (dateString) => {
  if (!dateString) return null;
  
  try {
    const date = new Date(dateString);
    // Create a new date with HK timezone components
    const hkString = date.toLocaleString(HK_LOCALE, { timeZone: HK_TIMEZONE });
    return new Date(hkString);
  } catch (error) {
    console.warn('Error converting to HK date:', error);
    return new Date(dateString);
  }
};

// Convenience functions for common use cases
export const hkNow = getHKNow;
export const hkNowISO = getHKNowISO;
export const hkDate = formatHKDate;
export const hkDateTime = formatHKDateTime;
export const hkDateOnly = formatHKDateOnly;
export const hkTimeOnly = formatHKTimeOnly;
export const hkRelative = getHKRelativeTime;
export const hkToday = isHKToday; 

File: backend/utils/timezone_utils.py
"""
Centralized Timezone Utilities for IQS Trade System
Ensures consistent Hong Kong time usage across the entire application
"""

from datetime import datetime, timedelta
import pytz
from typing import Optional, Union

# Hong Kong timezone
HK_TIMEZONE = pytz.timezone('Asia/Hong_Kong')

def get_hk_now() -> datetime:
    """
    Get current time in Hong Kong timezone.
    
    Returns:
        datetime: Current time in Hong Kong timezone
    """
    return datetime.now(HK_TIMEZONE)

def get_hk_now_iso() -> str:
    """
    Get current time in Hong Kong timezone as ISO format string.
    
    Returns:
        str: Current time in Hong Kong timezone as ISO format
    """
    return get_hk_now().isoformat()

def get_hk_now_str(format_str: str = '%Y-%m-%d %H:%M:%S') -> str:
    """
    Get current time in Hong Kong timezone as formatted string.
    
    Args:
        format_str: Format string for datetime.strftime()
    
    Returns:
        str: Current time in Hong Kong timezone as formatted string
    """
    return get_hk_now().strftime(format_str)

def get_hk_timestamp() -> str:
    """
    Get current time in Hong Kong timezone as timestamp string.
    
    Returns:
        str: Current time in Hong Kong timezone as YYYYMMDD_HHMMSS format
    """
    return get_hk_now().strftime('%Y%m%d_%H%M%S')

def convert_to_hk_time(dt: datetime) -> datetime:
    """
    Convert a datetime to Hong Kong timezone.
    If datetime is naive (no timezone), assumes UTC.
    
    Args:
        dt: datetime object to convert
    
    Returns:
        datetime: datetime in Hong Kong timezone
    """
    if dt.tzinfo is None:
        # Assume UTC if no timezone info
        dt = pytz.UTC.localize(dt)
    
    return dt.astimezone(HK_TIMEZONE)

def convert_to_hk_iso(dt: datetime) -> str:
    """
    Convert a datetime to Hong Kong timezone and return as ISO format.
    
    Args:
        dt: datetime object to convert
    
    Returns:
        str: datetime in Hong Kong timezone as ISO format
    """
    return convert_to_hk_time(dt).isoformat()

def get_hk_date_range(search_date_str: str) -> tuple[datetime, datetime]:
    """
    Get Hong Kong timezone date range for a given date string.
    
    Args:
        search_date_str: Date string in YYYY-MM-DD format
    
    Returns:
        tuple: (start_date, end_date) in Hong Kong timezone
    """
    search_date = datetime.strptime(search_date_str, '%Y-%m-%d')
    search_date = HK_TIMEZONE.localize(search_date)
    next_date = search_date + timedelta(days=1)
    return search_date, next_date

def get_hk_business_hours() -> dict:
    """
    Get Hong Kong business hours information.
    
    Returns:
        dict: Business hours information
    """
    return {
        'timezone': 'Asia/Hong_Kong',
        'business_hours': 'Monday to Friday, 9:00 AM to 5:00 PM (Hong Kong time)',
        'current_time': get_hk_now_str(),
        'is_business_hours': is_hk_business_hours()
    }

def is_hk_business_hours() -> bool:
    """
    Check if current time is within Hong Kong business hours.
    
    Returns:
        bool: True if within business hours (Mon-Fri, 9AM-5PM HK time)
    """
    now = get_hk_now()
    
    # Check if it's a weekday (Monday = 0, Sunday = 6)
    if now.weekday() >= 5:  # Saturday or Sunday
        return False
    
    # Check if it's between 9 AM and 5 PM
    business_start = now.replace(hour=9, minute=0, second=0, microsecond=0)
    business_end = now.replace(hour=17, minute=0, second=0, microsecond=0)
    
    return business_start <= now <= business_end

def get_hk_time_ago(hours: int = 0, days: int = 0, minutes: int = 0) -> datetime:
    """
    Get Hong Kong time from specified time ago.
    
    Args:
        hours: Number of hours ago
        days: Number of days ago
        minutes: Number of minutes ago
    
    Returns:
        datetime: Hong Kong time from specified time ago
    """
    now = get_hk_now()
    return now - timedelta(hours=hours, days=days, minutes=minutes)

def format_hk_time_for_display(dt: datetime, format_str: str = '%Y-%m-%d %H:%M:%S') -> str:
    """
    Format a datetime for display in Hong Kong timezone.
    
    Args:
        dt: datetime object to format
        format_str: Format string for strftime
    
    Returns:
        str: Formatted datetime string in Hong Kong timezone
    """
    hk_dt = convert_to_hk_time(dt)
    return hk_dt.strftime(format_str)

def get_hk_timezone_info() -> dict:
    """
    Get Hong Kong timezone information.
    
    Returns:
        dict: Timezone information
    """
    now = get_hk_now()
    return {
        'timezone': 'Asia/Hong_Kong',
        'current_time': now.isoformat(),
        'current_time_formatted': now.strftime('%Y-%m-%d %H:%M:%S'),
        'utc_offset': now.strftime('%z'),
        'is_dst': now.dst() != timedelta(0),
        'timezone_name': now.strftime('%Z')
    }

# Convenience functions for common use cases
def hk_now() -> datetime:
    """Alias for get_hk_now()"""
    return get_hk_now()

def hk_now_iso() -> str:
    """Alias for get_hk_now_iso()"""
    return get_hk_now_iso()

def hk_timestamp() -> str:
    """Alias for get_hk_timestamp()"""
    return get_hk_timestamp() 